orca-core/src/main/java/com/netflix/spinnaker/orca/locks/LockExtendingTaskExecutionInterceptor.java

```diff
@@ -0,0 +1,112 @@
+package com.netflix.spinnaker.orca.locks;
+
+import com.netflix.spinnaker.orca.ExecutionStatus;
+import com.netflix.spinnaker.orca.Task;
+import com.netflix.spinnaker.orca.TaskExecutionInterceptor;
+import com.netflix.spinnaker.orca.TaskResult;
+import com.netflix.spinnaker.orca.pipeline.AcquireLockStage;
+import com.netflix.spinnaker.orca.pipeline.ReleaseLockStage;
+import com.netflix.spinnaker.orca.pipeline.model.Stage;
+import com.netflix.spinnaker.orca.pipeline.tasks.AcquireLockTask;
+import com.netflix.spinnaker.orca.pipeline.tasks.ReleaseLockTask;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.time.Duration;
+import java.util.*;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+
+@Component
+public class LockExtendingTaskExecutionInterceptor implements TaskExecutionInterceptor {
+  private final LockManager lockManager;
+  private final LockingConfigurationProperties lockingConfigurationProperties;
+
+  @Autowired
+  public LockExtendingTaskExecutionInterceptor(LockManager lockManager,
+                                               LockingConfigurationProperties lockingConfigurationProperties) {
+    this.lockManager = lockManager;
+    this.lockingConfigurationProperties = lockingConfigurationProperties;
+  }
+
+  @Override
+  public long maxTaskBackoff() {
+    return Duration.ofSeconds(
+      lockingConfigurationProperties.getTtlSeconds() - lockingConfigurationProperties.getBackoffBufferSeconds()
+    ).toMillis();
+  }
+
+  @Override
+  public Stage beforeTaskExecution(Task task, Stage stage) {
+    extendLocks(task, stage);
+    return stage;
+  }
+
+  @Override
+  public TaskResult afterTaskExecution(Task task, Stage stage, TaskResult taskResult) {
+    extendLocks(task, stage);
+    return taskResult;
+  }
+
+  private void extendLocks(Task task, Stage stage) {
+    if (!lockingConfigurationProperties.isEnabled()) {
+      return;
+    }
+    if (task instanceof AcquireLockTask || task instanceof ReleaseLockTask) {
+      return;
+    }
+
+    Map<HeldLock, AtomicInteger> heldLocks = new HashMap<>();
+    Set<String> lockTypes = new HashSet<>(Arrays.asList(AcquireLockStage.PIPELINE_TYPE, ReleaseLockStage.PIPELINE_TYPE));
```

Comment:
If we're not against guava, this would work well here:

       Set<String> lockTypes = ImmutableSet.of(AcquireLockStage.PIPELINE_TYPE, ReleaseLockStage.PIPELINE_TYPE))